### PR TITLE
cmake: always prefer local symbols

### PR DESCRIPTION
**Pull request by tchaikov**
Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
@yaozongyou mind helping test this PR?

**Comment by yaozongyou** [comment]
the test `unittest_mutex` failed,  maybe it needs to be checked.

**Comment by tchaikov** [comment]
3251161 partially reverts #9940

File: src/test/common/test_mutex.cc [hunk 1]
```diff
@@ -41,12 +28,6 @@ static void disable_lockdep() {
   }
 }
 
-TEST(Mutex, NormalAsserts) {
```
Comment (yaozongyou) [inline src/test/common/test_mutex.cc]: We can keep this test by making the following change:
```cpp
+  testing::FLAGS_gtest_death_test_style = "threadsafe";
   Mutex* m = new Mutex("Normal",false);
   m->Lock();
-  EXPECT_THROW(m->Lock(), int);
+  EXPECT_DEATH(m->Lock(), ".*");
```

Comment (tchaikov) [reply src/test/common/test_mutex.cc]: your fix hangs.

Comment (tchaikov) [reply src/test/common/test_mutex.cc]: @yaozongyou fixed and repushed.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -452,6 +452,18 @@ target_link_libraries(ceph-common ${ceph_common_deps})
 set_target_properties(ceph-common PROPERTIES
   SOVERSION 0
   INSTALL_RPATH "")
+if(NOT APPLE)
+  # Apple uses Mach-O, not ELF. so this option does not apply to APPLE.
+  #
+  # prefer the local symbol definitions when binding references to global
+  # symbols. otherwise we could reference the symbols defined by the application
+  # with the same name, instead of using the one defined in libceph-common.
+  set_property(
```
Comment (aclamk) [inline src/CMakeLists.txt]: Maybe restate comment? Eg: "Require libceph-common to use local symbols, even if redefined in application".

Comment (tchaikov) [reply src/CMakeLists.txt]: sure, will do.

**Comment by tchaikov** [comment]
@aclamk updated and repushed.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-08-03_07:35:53-rados-wip-kefu-testing-2018-08-03-0109-distro-basic-smithi/

- some of the failures are caused by #23257
- the other are fixed by #23411

**Comment by andyzyao** [comment]
From jenkins console, core dumped:

```bash
2018-08-03 16:44:36.182 80f17b000 -1 *** Caught signal (Segmentation fault) **
 in thread 80f17b000 thread_name:

 ceph version Development (no_version) nautilus (dev)
 1: <handle_fatal_signal(int)+0xc8> at /home/jenkins/workspace/ceph-master/build/bin/ceph-mon
 2: <pthread_sigmask()+0x536> at /lib/libthr.so.3
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

--- begin dump of recent events ---
```

**Pull request merged**